docs/proposals/22-autoscaling-kube-apiserver-via-independent-hpa-and-vpa.md

```diff
@@ -0,0 +1,373 @@
+the use of sufficiently independent driving signals. HPA is driven by the rate of HTTP requests to kube-apiserver. VPA is
+driven by CPU and memory usage metrics.
+
+### Rationale
+A rough, heuristic horizontal sizing sets the stage for optimal VPA operation (horizontal size is within stable and
+reasonably efficient range). Vertical scaling fine-tunes requests, ensuring high efficiency (requests not too high)
+and stability (requests not too low).
+
+### Design Outline
+One HPA and one VPA resources are deployed alongside each ShootKapi deployment object. The VPA controls CPU and memory
+requests but not limits. The HPA is driven by the average rate of API requests per pod of the target kube-apiserver
+deployment.
+
+An API request rate metric is deployed to shoot prometheus (one per shoot). It gets aggregated by the aggregate
+prometheus of the respective seed (one per seed) through Gardener's existing metrics aggregation mechanism.
+
+A new component named prometheus-metrics-adapter is deployed on each seed. It sources performance metrics data from the
+seed aggregate prometheus and makes it available in the K8s metrics format. The prometheus-metrics-adapter is registered
+as an extension API service to the seed's kube-apiserver and takes responsibility of providing all custom metrics for
+that kube-apiserver.
+
+The request rate signal driving HPA originates in the shoot prometheus. It is propagated to the aggregate seed
+prometheus, and then further mediated by the prometheus-metrics-adapter and exposed as custom metric in the seed
+kube-apiserver. The custom metric in turn is directly consumed by the seed's HPA controller.
+
+![Design outline](./assets/gep22-design-outline.png "Design outline")
+
+_Fig 2: Proposed design_
+
+### Element: New Metric in Shoot Prometheus
+A new `shoot:apiserver_request_total:sum` recording rule is added to the shoot prometheus. To minimize latency, it
+calculates request rate across the minimum period for which two distinct samples are reliably available.
+
+### Element: Metric Propagation from Shoot to Seed
+To propagate metric data from individual shoot control planes to the seed level, the existing aggregate prometheus in
+the seed's garden namespace
+is leveraged. The aggregate prometheus already has access to shoot prometheus data, and implements a wildcard-based
+metric aggregation mechanism which, as long as certain naming convention is followed, aggregates the aforementioned new
+shoot prometheus metric without any need for changes.
+
+### Element: Prometheus Metrics Adapter
+A new component, named prometheus-metrics-adapter is added to seed clusters. It is an instance of the
+[kubernetes-sigs/prometheus-adapter]. It implements the
+Kubernetes Custom Metric API, exposing prometheus data in the form of K8s metrics selectively - controlled by
+configuration. A single replica is used, as highly available autoscaling setup is currently prevented by limitations in
+the autoscalers utilized by Gardener.
```

Comment:
Multiple VPA replicas don't play nicely, when acting on the same resources. Important, because Gardener aims for high availability, based on multiple component replicas. The note basically says we don't have that for the autoscaling setup.

Reply:
Marking this discussion thread resolved to keep the PR manageable. Please don't hesitate to reopen as appropriate.